lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -647,44 +647,38 @@ def retry_multiple(
         if not job_ids:
             raise ValueError("No job_ids provided to retry")
 
-        offending_ids = defaultdict(int)
-        for job_id in job_ids:
-            if job_ids.count(job_id) > 1:
-                offending_ids[job_id] += 1
-
-        if offending_ids.keys():
+        offending_ids = [item for item, count in Counter(job_ids).items() if count > 1]
+        if offending_ids:
             raise ValueError(
                 f"Retry of the same id in the same request is not supported."
-                f" Offending ids:{list(offending_ids.keys())} "
+                f" Offending ids:{offending_ids} "
             )
 
         # Check all inputs before attempting to start submitting jobs
         retried_jobs = []
-        jobs = []
-        batch_jobs = []
         for job_id in job_ids:
+            # Check for presubmission failures
             try:
                 job, batch_job = self._validate_retry_presubmit(
                     job_id=job_id, as_admin=as_admin
                 )
-                jobs.append(job)
-                batch_jobs.append(batch_job)
             except Exception as e:
-                raise RetryFailureException(e)
-
-        # Submit all of the collected jobs
-        for i, job_id in enumerate(job_ids):
+                # Collect the presubmit error and don't submit the job
+                retried_jobs.append({"job_id": job_id, "error": f"{e}"})
+                continue
```

Comment:
Would it make sense to log the stack trace here? We're completely losing the trace IIUC

Reply:
Depends if we can easily fix the error without the stack trace. If we can, we don't need it, if we need the stack trace, then we should include it. Do you need the stacktrace @ialarmedalien ?

Reply:
But the except catches all exceptions. How can you know if we can fix the error for any possible error that can occur?

Reply:
You don't, but if we can fix it, then we can, and we cannot, then we cannot. (We don't even document the types of possible exceptions that can be returned by the endpoint, and if we did, we would have to parse the string for errors?   )

Reply:
> You don't, but if we can fix it, then we can, and we cannot, then we cannot. 

how can you know if you can fix it or not without the stack trace? The stack trace may give you the information you need to diagnose or fix the error, especially for something like a `KeyError`.

> We don't even document the types of possible exceptions that can be returned by the endpoint

I'm confused here - what does this have to do with logging the exception?

Reply:
>how can you know if you can fix it or not without the stack trace? The stack trace may give you the information you need to diagnose or fix the error, especially for something like a KeyError.
Right, it would be hard for something like a KeyError

However, If the error message for the thrown exception is descriptive enough that you don't need a traceback, then you can fix it without the traceback. If it seems that most errors won't be descriptive enough then we may need a traceback.  There's plenty of APIs that return an error without a traceback. But maybe since our jsonRPC endpoints don't, we need the traceback for most of the errors?

>I'm confused here - what does this have to do with logging the exception?

It doesn't have to do with logging the exception, it has to do with fixing the exception as per your comment
> How can you know if we can fix the error for any possible error that can occur?

You can't know if you can fix the error or not. If its a system error, you can't fix it, and if its a user error, you might be able to fix it.

Reply:
> There's plenty of APIs that return an error without a traceback.

Hold on - are you talking about adding the traceback to the API return? I'm talking about logging it, just
```
self.logger.error(e)
```

Reply:
Yea, for some reason I thought by log you meant include in the API return, which makes sense to me if our errors messages are so bad that we don't know what the error is without the traceback.

Reply:
> Yea, for some reason I thought by log you meant include in the API return

Nopety nope nope

> our errors messages are so bad that we don't know what the error is without the traceback.

It's not even so much that, it's that when something unexpected happens or we have a bug (like a KeyError) then we're not writing the error messages, and so we're at the mercy of the authors of the code where it originated.

Also having line numbers both for our code and 3rd party code makes it much faster to fix bugs.

Reply:
I'd rather not pollute the logs with logging the (first exception) presubmit error, but writing to the db and submitting probably might be good candidates for logging (second exception).  Do you have strong feelings about logging both of them?

Reply:
Oh I see. Maybe we should probably have a general exception handler code that logs certain errors to the logs, and doesn't log others? And we should log the errors right when they occur, and not wait till they bubble up to log them?

Reply:
> I'd rather not pollute the logs with logging the (first exception) presubmit error, but writing to the db and submitting probably might be good candidates for logging (second exception). Do you have strong feelings about logging both of them?

The issue I have with not logging the 1st error is that if there's a bug in that code then it's going to make debugging stuff much harder. If the try / catch only caught errors where we know we don't need the exception that'd be fine, but right now it catches everything. Once potential option is making a custom exception subclass for the types of errors you care about and just catch that.

Reply:
> Oh I see. Maybe we should probably have a general exception handler logger code that logs certain errors to the logs, and doesn't log others? And we should log the errors right when they occur, and not wait till they bubble up to log them?

That all sounds fine to me. As long as we're not losing stack traces we need to debug issues in the field I'm happy.

Reply:
For now we could just log everything and add the improved code that's smarter about what gets logged later

Reply:
Hmm. Does that mean we would have to change every ValueError and bare exception we raise to be a subclass of ExecutionEngineException , and check for ExecutionEngineExceptions vs regular exceptions?

Reply:
I've been removing logging code, because I got tired of looking through logs that were useless so I'm hesitant to add more logging code but if you think thats what we need we can add in logger.error or logger.debug statements.

Reply:
> Does that mean we would have to change every ValueError ...

yep, just one suggestion, there may be other ways to do this

Reply:
> I've been removing logging code, because I got tired of looking through logs that were useless so I'm hesitant to add more logging code but if you think thats what we need we can add in logger.error or logger.debug statements.

yeah, I hear you there, but when ee2 starts throwing weird unexpected exceptions we'll start cursing if there's no way to get the exceptions.